src/Moryx.Runtime.Kernel/Modules/Components/ModuleDependencyManager.cs

```diff
@@ -85,130 +81,150 @@ private void BuildFacadeCache(IEnumerable<IServerModule> allServices)
+            foreach (var module in modules)
             {
-                var importingProperties = GetImportingProperties(service);
+                var importingProperties = GetImportingProperties(module);
                 foreach (var importingProperty in importingProperties)
                 {
                     var propType = importingProperty.Property.PropertyType;
                     if (typeof(Array).IsAssignableFrom(propType))
-                        FillArray(propType, importingProperty, service);
+                        FillArray(propType, importingProperty, module, facadeProviders);
                     else
-                        FillFacade(propType, importingProperty, service);
+                        FillFacade(propType, importingProperty, module, facadeProviders);
                 }
+
+                // Make sure no imports are missing
+                if (importingProperties.All(ip => !ip.Missing))
+                    satisfiedModules.Add(module);
             }
+
+            return satisfiedModules;
         }
 
-        private void FillArray(Type propType, ImportingProperty importingProperty, IServerModule service)
+        private void FillArray(Type propType, ImportingProperty importingProperty, IServerModule module, IDictionary<object, IServerModule> facadeProviders)
         {
             var elemType = propType.GetElementType();
-            var facades = _facadeCache.Keys.Where(elemType.IsInstanceOfType).ToArray();
+            var facades = facadeProviders.Keys.Where(elemType.IsInstanceOfType).ToArray();
             var facadeArray = Array.CreateInstance(elemType, facades.Length);
             for (int i = 0; i < facades.Length; i++)
             {
                 facadeArray.SetValue(facades[i], i);
             }
-            importingProperty.Property.SetValue(service, facadeArray);
+            importingProperty.Property.SetValue(module, facadeArray);
         }
 
 
-        private void FillFacade(Type propType, ImportingProperty importingProperty, IServerModule service)
+        private void FillFacade(Type propType, ImportingProperty importingProperty, IServerModule service, IDictionary<object, IServerModule> facadeProviders)
         {
-            var match = _facadeCache.Keys.FirstOrDefault(propType.IsInstanceOfType);
-            if (match != null || importingProperty.Attribute.IsOptional)
+            var match = facadeProviders.Keys.FirstOrDefault(propType.IsInstanceOfType);
+            if (match != null)
             {
-                if (match != null)
-                    importingProperty.Property.SetValue(service, match);
-                return;
+                importingProperty.Property.SetValue(service, match);
+            }
+            else if (importingProperty.Attribute.IsOptional)
+            {
+                _logger.Log(LogLevel.Warning, "Found no plugin hosting a facade of type {0} which is referenced by {1}.{2}",
```

Comment:
It is an old log message but would you change this to "module" insteaf of "plugin"?